src/Interval.h

```diff
@@ -0,0 +1,301 @@
+/*
+ * LSST Data Management System
+ * Copyright 2014-2015 AURA/LSST.
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <https://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#ifndef LSST_SG_INTERVAL_H_
+#define LSST_SG_INTERVAL_H_
+
+/// \file
+/// \author Serge Monkewitz
+/// \brief This file defines a template representing closed real intervals.
+
+#include <algorithm>
+#include <ostream>
+
+#include "SpatialRelation.h"
+
+
+namespace lsst {
+namespace sg {
+
+/// `Interval` represents a closed interval of the real numbers
+/// by its upper and lower bounds. It is parameterized by the Scalar
+/// type, which must be constructible from a double.
+///
+/// An interval with identical upper and lower bounds contains
+/// a single point, and is equal to that scalar bound.
+///
+/// An interval with an upper bound less than its lower bound is
+/// empty (contains no points), as are intervals with NaN as
+/// either bound.
+template <typename Derived, typename Scalar>
+class Interval {
+public:
+    /// This constructor creates an empty interval.
+    Interval() : _a(1.0), _b(0.0) {}
+    /// This constructor creates a closed interval containing only x.
+    explicit Interval(Scalar x) : _a(x), _b(x) {}
+    /// This constructor creates an interval from the given endpoints.
+    Interval(Scalar x, Scalar y) : _a(x), _b(y) {}
+
+    /// Two closed intervals are equal if their endpoints are the same, or
+    /// both are empty.
+    bool operator==(Interval const & i) const {
+        return (_a == i._a && _b == i._b) || (i.isEmpty() && isEmpty());
+    }
+    bool operator!=(Interval const & i) const { return !(*this == i); }
+
+    /// A closed interval is equal to a point x if both endpoints equal x.
+    bool operator==(Scalar x) const {
+        return (_a == x && _b == x) || (x != x && isEmpty());
+    }
+    bool operator!=(Scalar x) const { return !(*this == x); }
+
+    ///@{
+    /// `a` returns the lower endpoint of this interval. The return value
+    /// for empty intervals is arbitrary.
+    Scalar a() const { return _a; }
+    Scalar & a() { return _a; }
+    ///@}
```

Comment:
These accessors (and the ones below for `b`, `center`, and `size`) don't conform to our coding standards, which require `getA`, `getB`, `getCenter`, `getSize`.

Also, if you have an explicit need for a non-const reference accessor, fine - but I think a setter is preferable if not, or perhaps even just leaving the class immutable aside from the assignment operator.  I've frequently found that for simple classes it can turn out to be convenient for all mutation to go through the constructor, especially if the class is ever exposed to Python.